### DISCUSSION TITLE
v2.2.0 - 2024-03-29

### DISCUSSION THREAD

**Discussion by gsnider2195** [nautobot/nautobot#5510] [General]
#### Contact and Team Models ([#230](https://github.com/nautobot/nautobot/issues/230))

Contact and Team are models that represent an individual and a group of individuals who can be linked to an object. Contacts and teams store the necessary information (name, phone number, email, and address) to uniquely identify and contact them. They are added to track ownerships of organizational entities and to manage resources more efficiently in Nautobot. Check out the documentation for [Contact](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/extras/contact/) and [Team](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/extras/team/). There is also a [user guide](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/feature-guides/contacts-and-teams/) available on how to utilize these models.

A new management command has been introduced to assist with migrating the Location fields `contact_name`, `contact_phone` and `contact_email` to the new Contact and Team models. This command can be invoked with `nautobot-server migrate_location_contacts` and will present a series of prompts to guide you through migrating Locations that have data in the `contact_name`, `contact_phone`, or `contact_email` fields which are not already associated to a Contact or Team. This command will give you the option to create new Contacts or Teams or, if a similar Contact or Team already exists, to link the Location to the existing Contact or Team. Note that when assigning a Location to an existing Contact or Team that has a blank `phone` or `email` field, the value from the Location will be copied to the Contact/Team. After a Location has been associated to a Contact or Team, the `contact_name`, `contact_phone`, and `contact_email` fields will be cleared from the Location.

#### Controller Model ([#3111](https://github.com/nautobot/nautobot/issues/3111))

Controller models have been added to the `dcim` app. A Controller in Nautobot is an abstraction meant to represent network or SDN (Software-Defined Networking) controllers. These may include, but are not limited to, wireless controllers, cloud-based network management systems, and other forms of central network control mechanisms.

For more details, refer to the user guide for a [`Controller` model](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/controller/), a [`ControllerManagedDeviceGroup` model](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/controllermanageddevicegroup/), or developer documentation for [Controllers](https://docs.nautobot.com/projects/core/en/v2.2.0/development/core/controllers/).

#### DeviceFamily Model ([#3559](https://github.com/nautobot/nautobot/issues/3559))

A [Device Family](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/devicefamily/) represents a group of related [Device Types](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/devicetype/). A Device Type can be optionally assigned to a Device Family. Each Device Family must have a unique name and may have a description assigned to it.

#### Jobs Tile View ([#5129](https://github.com/nautobot/nautobot/issues/5129))

Job list is now available in two display variants: list and tiles. List is a standard table view with no major changes introduced. Tiles is a new type of view displaying jobs in a two-dimensional grid.

#### Prefix and VLAN Many Locations ([#4334](https://github.com/nautobot/nautobot/issues/4334), [#4412](https://github.com/nautobot/nautobot/issues/4412))

The Prefix and VLAN models have replaced their single `location` foreign-key field with a many-to-many `locations` field, allowing multiple Locations to be attached to a single Prefix or VLAN. To ensure backwards compatibility with pre-2.2 code, these models now have a `location` property which can be retrieved or set for the case of a single associated Location, but will raise a `MultipleObjectsReturned` exception if the Prefix or VLAN in question has more than one associated Location. REST API versions 2.0 and 2.1 similarly still have a `location` field, while REST API version 2.2 and later replace this with `locations`.

#### Software Image File and Software Version models ([#1](https://github.com/nautobot/nautobot/issues/1))

New models have been added for Software Image Files and Software Versions. These models are used to track the software versions of Devices, Inventory Items and Virtual Machines and their associated image files. These models have been ported from the [Device Lifecycle Management App](https://github.com/nautobot/nautobot-app-device-lifecycle-mgmt/) and a future update to that app will migrate all existing data from the `nautobot_device_lifecycle_mgmt.SoftwareImageLCM` and `nautobot_device_lifecycle_mgmt.SoftwareLCM` models to the `dcim.SoftwareImageFile` and `dcim.SoftwareVersion` models added here.

Software Versions must be associated to a Platform. Software Image Files must be associated to one Software Version and may be associated to one or more Device Types. Devices, Inventory Items and Virtual Machines may be associated to one Software Version to track their current version. See the documentation for [Software Image File](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/softwareimagefile/) and [Software Version](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/core-data-model/dcim/softwareversion/). There is also a [user guide](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/feature-guides/software-image-files-and-versions/) with instructions on how to create these models.

#### Syntax Highlighting ([#5098](https://github.com/nautobot/nautobot/issues/5098))

Language syntax highlighting for GraphQL, JSON, XML and YAML is now supported in the UI via JavaScript. To enable the feature, a code snippet has to be wrapped in the following HTML structure:

```html
<pre><code class="language-{graphql,json,xml,yaml}">...</code></pre>
```

[`render_json`](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/platform-functionality/template-filters/#render_json) and [`render_yaml`](https://docs.nautobot.com/projects/core/en/v2.2.0/user-guide/platform-functionality/template-filters/#render_yaml) template filters default to this new behavior with an optional opt-out `syntax_highlight=False` arg.

#### Data Imports as a System Job ([#5064](https://github.com/nautobot/nautobot/issues/5064))

The CSV import functionality for all models has been changed from a synchronous operation to an asynchronous background task (system Job). As a result, imports of large CSV files will no longer fail due to browser timeout.

!!! tip
    Users now must have the `run` action permission for `extras > job` (specifically the `nautobot.core.jobs.ImportObjects` Job) in order to import objects, in addition to the normal `add` permissions for the object type being imported.

#### Plugin to App Renames([#5341](https://github.com/nautobot/nautobot/issues/5341))

`Installed Plugins` view has been renamed to `Installed Apps`. `Plugin` terminologies in `Installed Plugins` (now `Installed Apps`) view and dependent views have been changed to `App` throughout. `Plugin` references in documentation (excluding old release-notes) have been replaced by `App`. `Plugins` navigation menu has been renamed to `Apps`.

#### Standardization of `max_length` on all Charfields ([#2906](https://github.com/nautobot/nautobot/issues/2906))

Model CharFields' `max_length` attributes have been standardized globally to have at least 255 characters except where a shorter `max_length` is explicitly justified.

**Full Changelog**: https://github.com/nautobot/nautobot/compare/v2.1.9...v2.2.0

## Contributors

* @bradh11 
* @snaselj
* @edwardtheharris
* @gsnider2195
* @timizuoebideri1
* @abates
* @mzbroch
* @glennmatthews
* @lampwins
* @HanlinMiao
* @whitej6
* @norbert-mieczkowski-codilime 
* @gertzakis 

<hr /><em>This discussion was created from the release <a href='https://github.com/nautobot/nautobot/releases/tag/v2.2.0'>v2.2.0 - 2024-03-29</a>.</em>

**Comment by bradh** [comment]
I'm not sure I have any contribution here. Possibly @bradh11 ?

**Reply by gsnider2195** [reply]
Sorry about that. Must have been a copy/paste error.
